### PR TITLE
Java: add Android database taint and SQL injection sinks

### PR DESCRIPTION
**Pull request by aibaars**
This pull request adds taint steps and SQL injection sinks for the following classes of the Android database library: `DatabaseUtils`, `SQLiteDatabase`, `SQLiteQueryBuilder`, and part of `ContentProvider/Resolver`.

The [SQLiteQueryBuilder](https://github.com/aosp-mirror/platform_frameworks_base/blob/master/core/java/android/database/sqlite/SQLiteQueryBuilder.java) class mostly performs unverified string concatenation for any parts of the SQL query string. This can be mitigated by setting its strictness flags which harden the query builder against SQL injection attempts. This pull request does not model the strictness flags and can therefore lead to false positives. We can improve that in the future if the false positive rate is deemed too high.

### CODE REVIEW

File: java/ql/src/semmle/code/java/frameworks/android/SQLite.qll [hunk 1]
```diff
@@ -1,45 +1,219 @@
 import java
+import Android
 
+/**
+ * The class android.database.sqlite.SQLiteDatabase
```
Comment (intrigus-lgtm) [inline java/ql/src/semmle/code/java/frameworks/android/SQLite.qll:5]:
```suggestion
 * The class `android.database.sqlite.SQLiteDatabase`
```

File: java/ql/src/semmle/code/java/frameworks/android/SQLite.qll [hunk 2]
```diff
@@ -1,45 +1,219 @@
...
+ */
 class TypeSQLiteDatabase extends Class {
   TypeSQLiteDatabase() { hasQualifiedName("android.database.sqlite", "SQLiteDatabase") }
 }
 
+/**
+ * android.database.sqlite.SQLiteQueryBuilder
```
Comment (intrigus-lgtm) [inline java/ql/src/semmle/code/java/frameworks/android/SQLite.qll:12]:
```suggestion
 * The class `android.database.sqlite.SQLiteQueryBuilder`
```

File: java/ql/src/semmle/code/java/frameworks/android/SQLite.qll [hunk 3]
```diff
@@ -1,45 +1,219 @@
...
+ */
+class TypeSQLiteQueryBuilder extends Class {
+  TypeSQLiteQueryBuilder() { hasQualifiedName("android.database.sqlite", "SQLiteQueryBuilder") }
+}
+
+/**
+ * The class android.database.sqlite.DatabaseUtils
```
Comment (intrigus-lgtm) [inline java/ql/src/semmle/code/java/frameworks/android/SQLite.qll:19]:
```suggestion
 * The class `android.database.sqlite.DatabaseUtils`
```

File: java/ql/src/semmle/code/java/frameworks/android/SQLite.qll [hunk 4]
```diff
@@ -1,45 +1,228 @@
...
+  QueryBuilderQueryMethod() {
+    // query(SQLiteDatabase db, String[] projectionIn, String selection, String[] selectionArgs, String groupBy, String having, String sortOrder)
+    // query(SQLiteDatabase db, String[] projectionIn, String selection, String[] selectionArgs, String groupBy, String having, String sortOrder, String limit)
+    // query(SQLiteDatabase db, String[] projectionIn, String selection, String[] selectionArgs, String groupBy, String having, String sortOrder, String limit, CancellationSignal cancellationSignal)
+    this.getDeclaringType().getASourceSupertype*() instanceof TypeSQLiteQueryBuilder and
+    this.hasName("query")
+  }
+
+  override int sqlIndex() { result = [-1, 3, 5, 6, 7, 8] }
+}
```
Comment (joefarebrother) [inline java/ql/src/semmle/code/java/frameworks/android/SQLite.qll:175]: These indicies look off by one - the `selection` arg is at index 2

**Review by joefarebrother — APPROVED** [review]
Looks good

**Review by aschackmull — APPROVED** [review]

**Review by aschackmull — APPROVED** [review]

**Pull request merged**